### PR TITLE
Modify rule S4335: Adapt to LaYC

### PR DESCRIPTION
**Pull request by victor-diez-sonarsource**

### CODE REVIEW

File: rules/S4335/javascript/metadata.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "title": "Types without members, 'any' and 'never' should not be used in type intersections",
+  "title": "Type intersections should use meaningful types",
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/metadata.json:2]:
```suggestion
  "title": "Type intersections should compose consistent types",
```

Comment (yassin-kammoun-sonarsource) [reply rules/S4335/javascript/metadata.json:2]:
```suggestion
  "title": "Type intersections should compose types consistently",
```

Comment (yassin-kammoun-sonarsource) [reply rules/S4335/javascript/metadata.json:2]:
```suggestion
  "title": "Type intersections should use consistent type constituents",
```

Comment (yassin-kammoun-sonarsource) [reply rules/S4335/javascript/metadata.json:2]:
```suggestion
  "title": "Type intersections should define consistent type constituents",
```

Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/metadata.json:2]:
```suggestion
  "title": "Type intersections should use consistent types",
```

Comment (yassin-kammoun-sonarsource) [reply rules/S4335/javascript/metadata.json:2]: I can't make up my mind :/

File: rules/S4335/javascript/rule.adoc [hunk 2]
```diff
@@ -1,39 +1,47 @@
 == Why is this an issue?
 
-An intersection type combines multiple types into one. This allows you to add together existing types to get a single type that has all the features you need. However an intersection with a type without members doesn't change the resulting type. In the opposite the usage of ``++any++`` or ``++never++`` as part of an intersection will always results in ``++any++`` or ``++never++`` respectively. This is almost certainly an error.
+In TypeScript, intersections are used to combine multiple types into a single one, allowing you to create more flexible and powerful type definitions. An intersection type is represented using the ampersand symbol ``++&++``. It allows you to combine multiple types into a single type that includes all the properties and methods from each type.
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc:3]:
```suggestion
In TypeScript, intersections are used to combine multiple types into a single one. An intersection type is represented using the ampersand symbol ``++&++``. It allows you to combine multiple types into a single type that includes all the properties and methods from each type, thus creating more flexible and powerful type definitions.
```

Comment (yassin-kammoun-sonarsource) [reply rules/S4335/javascript/rule.adoc:3]: Let's avoid saying what it "allows" twice :)

File: rules/S4335/javascript/rule.adoc [hunk 3]
```diff
@@ -1,39 +1,47 @@
...
 
-=== Noncompliant code example
+* The `never` type represents the type of values that never occur. Intersecting any type with `never` will always result in type `never`.
 
-[source,javascript]
-----
-function foo(p: MyType & null) { // Noncompliant
- // ...
-}
+* The `any` type allows to opt-out of type checking during compilation. Variables of type any allow you to access arbitrary properties, even ones that don't exist. `any` comes at the cost of losing type safety, which is one of the main motivations for using TypeScript. Avoid using `any` when not necessary. Intersecting any type with `any` will always result in type `any`.
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc:9]:
```suggestion
* The `any` type allows to opt-out of type checking during compilation. Expressions of type `any` allow you to access arbitrary properties, even ones that don't exist. `any` comes at the cost of losing type safety, which is one of the main motivations for using TypeScript. Avoid using `any` when not necessary. Intersecting any type with `any` will always result in type `any`.
```

File: rules/S4335/javascript/rule.adoc [hunk 4]
```diff
@@ -1,39 +1,47 @@
...
+
+* The `undefined` and `null` are the types for their respective values. Intersecting any type with `never` will always result in type `never`.
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc]:
```suggestion
* The `undefined` and `null` types are the types for their respective value. Intersecting any type with `never` will always result in type `never`.
```

Comment (yassin-kammoun-sonarsource) [reply rules/S4335/javascript/rule.adoc]: Something is weird with the second sentence. Probably copy-pasted :/

File: rules/S4335/javascript/rule.adoc [hunk 5]
```diff
@@ -1,39 +1,47 @@
...
+
+* The `void` type implies the absence of a type. Intersecting any type with `never` will always result in type `never`.
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc]: Same here.

File: rules/S4335/javascript/rule.adoc [hunk 6]
```diff
@@ -1,39 +1,47 @@
...
 
-function bar(p: MyType & any) { // Noncompliant
- // ...
-}
+Additionally, an intersection with a type without members (for example, ``++{}++``) doesn't change the resulting type, is redundant, and can be safely removed from the intersection.
+
+[source,javascript,diff-id=1,diff-type=noncompliant]
 ----
+type MyType = MyOtherType & null; //Noncompliant: MyType is always 'never'
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc:19]:
```suggestion
type Foo = T & null; // Noncompliant: 'Foo' is always 'never'
```

File: rules/S4335/javascript/rule.adoc [hunk 7]
```diff
@@ -1,39 +1,47 @@
...
 
+type MyType = MyOtherType & any; //Noncompliant: MyType is always 'any'
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc:21]:
```suggestion
type Bar = T & any; // Noncompliant: 'Bar' is always 'any'
```

File: rules/S4335/javascript/rule.adoc [hunk 8]
```diff
@@ -1,39 +1,47 @@
...
 
-=== Compliant solution
+type MyType = MyOtherType & AnotherType & {}; //Noncompliant: '{}' has no memebers and is redundant
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc:23]:
```suggestion
type Baz = T & U & {}; // Noncompliant: '{}' has no members and is redundant
```

File: rules/S4335/javascript/rule.adoc [hunk 9]
```diff
@@ -1,39 +1,47 @@
...
-}
-
-function bar(p: any) {
- // ...
-}
+type MyType = MyOtherType | null;
+
+type MyType = any;
+
+type MyType = MyOtherType & AnotherType;
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc:34]:
```suggestion
type Baz = T & U;
```

File: rules/S4335/javascript/rule.adoc [hunk 10]
```diff
@@ -1,39 +1,47 @@
...
-function foo(p: MyType & AnotherType) {
- // ...
-}
-
-function bar(p: any) {
- // ...
-}
+type MyType = MyOtherType | null;
+
+type MyType = any;
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc:32]:
```suggestion
type Bar = any;
```

File: rules/S4335/javascript/rule.adoc [hunk 11]
```diff
@@ -1,39 +1,47 @@
...
-}
-// or
-function foo(p: MyType & AnotherType) {
- // ...
-}
-
-function bar(p: any) {
- // ...
-}
+type MyType = MyOtherType | null;
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc:30]:
```suggestion
type Foo = T | null;
```

File: rules/S4335/javascript/rule.adoc [hunk 12]
```diff
@@ -1,39 +1,47 @@
...
+type MyType = MyOtherType & null; //Noncompliant: MyType is always 'never'
 
+type MyType = MyOtherType & any; //Noncompliant: MyType is always 'any'
 
-=== Compliant solution
+type MyType = MyOtherType & AnotherType & {}; //Noncompliant: '{}' has no memebers and is redundant
+----
 
-[source,javascript]
+Use specific types that accurately represent the data being used.
```
Comment (yassin-kammoun-sonarsource) [inline rules/S4335/javascript/rule.adoc:26]:
```suggestion
Use consistent types that accurately reflect and restrict the domain of values of the defined data type.
```

**Review by yassin-kammoun-sonarsource — APPROVED** [review]

File: rules/S4335/javascript/rule.adoc (hunk 12 above)
Comment (victor-diez-sonarsource) [reply rules/S4335/javascript/rule.adoc:26]: I would not say restrict, intersections in principle are used to extend, not restrict, right?

Comment (yassin-kammoun-sonarsource) [reply rules/S4335/javascript/rule.adoc:26]: Yeah, but if you use `any`, you kinda open the gate to everything, right ?

Comment (yassin-kammoun-sonarsource) [reply rules/S4335/javascript/rule.adoc:26]: Maybe restrict is too strong. What about limit or narrow?

Comment (victor-diez-sonarsource) [reply rules/S4335/javascript/rule.adoc:26]: but intersecting with any raises an issue. I changed the compliant version.

what about 

```suggestion
Use consistent types that accurately reflect the domain of values of the defined data type.
```

File: rules/S4335/javascript/metadata.json (hunk 1 above)
Comment (victor-diez-sonarsource) [reply rules/S4335/javascript/metadata.json:2]: me neither

Comment (victor-diez-sonarsource) [reply rules/S4335/javascript/metadata.json:2]: not convinved about the word `consistent`

Comment (victor-diez-sonarsource) [reply rules/S4335/javascript/metadata.json:2]: what about `significant`?

File: rules/S4335/javascript/rule.adoc (hunk 12 above)
Comment (yassin-kammoun-sonarsource) [reply rules/S4335/javascript/rule.adoc:26]: Looks good

**Pull request merged**